Add snapshotter and differ for block CIMs

ambarve opened the pull request:
This commit adds the snapshotter and differ plugins that can be used to pull/import container images in the block CIM format. (More about block CIMs [here](https://github.com/microsoft/hcsshim/blob/main/pkg/cimfs/doc.go).)

kiashok commented on plugins/snapshots/windows/block_cimfs.go:
Can we add a Cleanup() function like the other linux snapshotters have to take care of deleting any orphaned directories? I think we had decided on adding one for windows snapshotter for long term too.

kiashok replied:
Sure we can do that.

kiashok commented:
LGTM

cpuguy83 reviewed (changes requested):
I don't see anything glaringly wrong here.

ambarve commented on plugins/snapshots/windows/block_cimfs.go:
```diff
@@ -0,0 +1,555 @@
+			}
+		}()
+
+		if strings.Contains(key, snapshots.UnpackKeyPrefix) {
```
Updated to use `HasPrefix`.

ambarve commented:
@fuweid & @cpuguy83  could you please take another look and/or approve this PR?

ambarve committed:
Block CIM snapshotter & differ

Add new snapshotter & differ plugins that can leverage the block CIM layer writers to
store pulled/imported images in the block CIM format.

Signed-off-by: Amit Barve <ambarve@microsoft.com>

slonopotamus commented on plugins/snapshots/windows/cimfs.go:
```diff
@@ -47,12 +47,48 @@ import (
-	vhdBlockSizeInBytes uint32 = 1 * 1024 * 1024         // 1 MB
+	baseVHDName                         = "blank-base.vhdx"
+	templateVHDName                     = "blank.vhdx"
+	defaultScratchVHDSizeInBytes uint64 = 10 * 1024 * 1024 * 1024 // 10 GB
```
Is it configurable? Why it is 10GB instead of 20GB (Hyper-V default for VHD)?

ambarve replied:
Yeah, this can be configured with the `func WithSize(size uint64) scratchCreationOpt` option. I don't remember if there was a specific reason to keep default of 10.
